# ExcisionFinder

ExcisionFinder is a tool for designing personalized and allele-specific sgRNAs for general genome editing and optimizing pairs of sgRNAs for human therapeutic editing.

**Note**: ExcisionFinder is currently in early beta and has many bugs, which we are actively working to fix. If you use this tool and find any issues or anything that is difficult to interpret, please send a message to this github or to kathleen.keough@gladstone.ucsf.edu

## Table of Contents

1. Getting started
* [How it works](https://github.com/keoughkath/ExcisionFinder/wiki/Overview)
* [Install ExcisionFinder and its requirements](https://github.com/keoughkath/ExcisionFinder/wiki/Install-ExcisionFinder-and-its-requirements)
2. Tool descriptions
* [Design allele-specific sgRNAs](https://github.com/keoughkath/ExcisionFinder/wiki/Usage:-gen_sgRNAs.py)
* [Using CRISPOR with ExcisionFinder](https://github.com/keoughkath/ExcisionFinder/wiki/Using-CRISPOR-with-gen_sgRNAs.py)
3. Tutorials
* [Tutorial: Design sgRNAs for allele specific excision of the gene MFN2 in the WTC genome](https://github.com/keoughkath/ExcisionFinder/wiki/Tutorial:-Design-sgRNAs-for-allele-specific-excision-of-the-gene-MFN2-in-the-WTC-genome)

# ExcisionFinder 1000 Genomes analysis dataset

The annotated variant database and gene targetability results from the 1000 Genomes Project are available [here](http://lighthouse.ucsf.edu/public_files_no_password/excisionFinderData_public/ExcisionFinder_manuscript_data/).

## Useful links

[Pre-computed PAM locations.](http://lighthouse.ucsf.edu/public_files_no_password/excisionFinderData_public/)

[UCSC chromosome fastas, hg19.](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/chromosomes/)

[UCSC chromosome fastas, hg38.](http://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/)


