# ExcisionFinder

ExcisionFinder is a database and toolset for analysis and design of allele-specific CRISPR editing that leverages >2,500 human genomes to estimate feasibility of allele-specific loci in varied individuals and design allele-specific sgRNAs in individual genomes. 

**Note**: ExcisionFinder is currently in early beta and has many bugs, which we are actively working to fix. If you use this tool and find any issues or anything that is difficult to interpret, please send a message to this github or to kathleen.keough@gladstone.ucsf.edu

## Table of Contents

1. Getting started
* [How it works](https://github.com/keoughkath/ExcisionFinder/wiki/Overview)
* [Install ExcisionFinder and its requirements](https://github.com/keoughkath/ExcisionFinder/wiki/Install-ExcisionFinder-and-its-requirements)
2. Tool descriptions
* [Design allele-specific sgRNAs](https://github.com/keoughkath/ExcisionFinder/wiki/Usage:-gen_sgRNAs.py)
* [Using CRISPOR with ExcisionFinder](https://github.com/keoughkath/ExcisionFinder/wiki/Using-CRISPOR-with-gen_sgRNAs.py)
3. Tutorials
* [Tutorial: Design sgRNAs for allele specific excision of the gene MFN2 in the WTC genome](https://github.com/keoughkath/ExcisionFinder/wiki/Tutorial:-Design-sgRNAs-for-allele-specific-excision-of-the-gene-MFN2-in-the-WTC-genome)

# ExcisionFinder 1000 Genomes analysis dataset

The annotated variant database and gene targetability results from the 1000 Genomes Project are available [here](http://lighthouse.ucsf.edu/public_files_no_password/excisionFinderData_public/ExcisionFinder_manuscript_data/).
