# ExcisionFinder

ExcisionFinder is a database and toolset for analysis and design of allele-specific CRISPR editing that leverages >2,500 human genomes to estimate feasibility of allele-specific loci in varied individuals and design allele-specific sgRNAs in individual genomes. 

## Table of Contents
1. Getting started
* [How it works](https://github.com/keoughkath/ExcisionFinder/wiki/Overview)
* [Install ExcisionFinder and its requirements](https://github.com/keoughkath/ExcisionFinder/wiki/Install-ExcisionFinder-and-its-requirements)
2. Reference database
* [Download default 1000 Genomes-based database](https://github.com/keoughkath/ExcisionFinder/wiki/Download-the-ExcisionFinder-database)
* [Query the database](https://github.com/keoughkath/ExcisionFinder/wiki/Querying-ExcisionFinder-variant-annotations)
3. Use ExcisionFinder to annotate a custom genome
4. Use ExcisionFinder to annotate a custom cohort
5. Step-by-step tutorials
* [Design allele-specific sgRNAs](https://github.com/keoughkath/ExcisionFinder/wiki/Tutorial:-design-sgRNAs-in-an-individual-genome)

# ExcisionFinder data

The annotated variant database and gene targetability results from the 1000 Genomes Project are available [here](http://lighthouse.ucsf.edu/public_files_no_password/excisionFinderData_public/ExcisionFinder_manuscript_data/).
